src/components/createContainer.jsx

```diff
@@ -32,41 +34,47 @@ export default function createContainer(DecoratedComponent, specs) {
     static DecoratedComponent = DecoratedComponent
 
     static contextTypes = {
-      renderLoading: PropTypes.func.isRequired,
-      adaptor: adaptorShape.isRequired,
+      adrenaline: PropTypes.shape({
+          renderLoading: PropTypes.func.isRequired,
+          adaptor: adaptorShape.isRequired
+      }).isRequired
+    }
+
+    static childContextTypes = {
+        adrenaline: containerShape.isRequired
     }
 
     constructor(props, context) {
       super(props, context);
-
       this.isNotMounted = true;
+      this.state = {};
+    }
 
-      const { adaptor } = context;
-      //DecoratedComponent.prototype.shouldComponentUpdate = adaptor.shouldComponentUpdate;
-
-      const initialArgs = !!specs.args ? specs.args(props) : {};
-
-      this.state = {
-        data: {},
-        args: initialArgs,
-      };
-
-      const dataKeys = Object.keys(specs.queries(initialArgs));
-      const { propTypes } = DecoratedComponent;
-      const requiredKeys = dataKeys.filter(key =>
-        propTypes[key] && !propTypes[key].isRequired
-      );
-
-      this.isDataLoaded = (data = {}) => requiredKeys.every(key => {
-        return data[key] !== null && data[key] !== undefined
-      });
+    getChildContext() {
+        const { adrenaline } = this.context;
+        const container = {
+            setArgs: this.setArgs,
+            state: this.state
+        };
+        return {
+            adrenaline: {
+                ...adrenaline,
+                container
+            }
+        };
     }
 
     componentDidMount() {
-      const { adaptor } = this.context;
+      const adaptor = this.getAdaptor();
       this.isNotMounted= false;
+      this.setArgs(this.computeArgs(this.props));
```

Comment:
I think this can be safely moved to ```componentWillMount()```. What do you think? The earlier we start loading data, the better. It could work in the constructor, but that comes with the risk of performing queries for something that may not actually be mounted.